Assets/Scripts/UI/EventLog.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using TMPro;
+
+internal struct EventLogItem
+{
+    public TMP_Text text;
+    public float creationTime;
+}
+
+public class EventLog : MonoBehaviour
+{
+    [SerializeField]
+    private RectTransform itemHolder;
+
+    [SerializeField]
+    private TMP_Text[] items;
+
+    [SerializeField]
+    private float expiryTime = 5;
+
+    private Queue<TMP_Text> availableItems = new();
+    private Queue<EventLogItem> itemsInUse = new();
+
+    private int currentIndex = 0;
+
+    public static EventLog Singleton { get; private set; }
+
+    private void Awake()
+    {
+        #region Singleton boilerplate
+
+        if (Singleton != null)
+        {
+            if (Singleton != this)
+            {
+                Debug.LogWarning($"There's more than one {Singleton.GetType()} in the scene!");
+                Destroy(gameObject);
+            }
+
+            return;
+        }
+
+        Singleton = this;
+
+        #endregion Singleton boilerplate
+
+        DontDestroyOnLoad(gameObject);
+    }
+
+    private void Start()
+    {
+        foreach (var item in items)
+        {
+            item.enabled = false;
+            availableItems.Enqueue(item);
+        }
+    }
+
+    public void Log(string message)
+    {
+        TMP_Text item;
+        if (availableItems.Count > 0)
+            item = availableItems.Dequeue();
+        else
+            item = itemsInUse.Dequeue().text;
+
+        item.text = message;
+        item.enabled = true;
+        item.transform.SetAsFirstSibling();
+
+        itemsInUse.Enqueue(new EventLogItem()
+        {
+            creationTime = Time.time,
+            text = item,
+        });
+    }
+
+    private void Update()
+    {
+        while (itemsInUse.Count > 0 && Time.time - itemsInUse.Peek().creationTime > expiryTime)
+        {
+            var item = itemsInUse.Dequeue();
+            // Fade and put back in queue after tween is done.
+            LeanTween.value(item.text.gameObject, SetAlpha(item.text), 1, 0, .2f).setOnComplete(() =>
+            {
+                item.text.enabled = false;
+                item.text.transform.SetAsLastSibling();
+                availableItems.Enqueue(item.text);
+            });
```

Comment:
There seems to be a bug where if enough items has been used, typically in the third shooting round, messages will no longer appear visually. I think it might be due to some of these shenaningans, specifically I think you need tween cancel guards that always properly resets the alpha of the canceled item so it's visible again the next time it's in use.

Reply:
Uh oh, so that did turn out to be necessary -_-

Reply:
The only required part of the fix was the `alpha = 1` reset when a new item is logged, but I went ahead and fixed the issue that *could* appear if we were to exceed the capacity of the log (I originally wrote it so that items were temporarily outside both queues while they faded out)